### ISSUE TITLE
Markdown: Support using `**`/`__`/`==` on selections

### ISSUE THREAD

**Issue by arv-anshul** [zed-industries/zed#12469]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Hello,

I have encountered an issue with the Zed code editor when trying to add ** around a selected word to bold it. When I select a word and attempt to add ** at the start and end in one go, the editor deletes the selected word instead. This behavior does not occur when using brackets; adding brackets around a selected word works as expected.

Steps to reproduce:
1. Select a word in the editor.
2. Attempt to add `**`/`__`/`==`/`` ` `` at the start and end of the selected word simultaneously.

Expected behavior:
The selected word should be wrapped in `**` and remain intact.

Actual behavior:
The selected word is deleted.

This issue affects the efficiency of formatting text in markdown or similar syntaxes where bolding text is required.

Thank you for looking into this!

Best regards,
Anshul Raj Verma

> Used ChatGPT to generate this. Also modified some bit.

### Environment

Zed: v0.137.3 (Zed)
OS: macOS 14.5.0
Memory: 16 GiB
Architecture: aarch64

**Comment by wagslane** [comment]
I've also run into this - I write a *ton* of markdown into Zed, and having this work would be incredible

**Comment by xpe** [comment]
@arv-anshul Thank you for filing this issue; otherwise I would have had to do it myself. :)

This issue to a large degree is still present in my version of Zed Preview: d4f44c1137fa5515a232c392d42efd69041717f9.

I've found that wrapping a selection in ` does indeed work. 

I would suggest also including the `~~` Markdown strikethrough characters. For example: ~~Begone stale text!~~

**Comment by xpe** [comment]
I'm browsing the source code ([crates/markdown/src/parser.rs](https://github.com/zed-industries/zed/blob/main/crates/markdown/src/parser.rs)) to assess how much work would be involved. Some notes:

These span-level tags are relevant:
- `MarkdownTag::Strong` for `**` and `__` (bold)
- `MarkdownTag::Emphasis` for `*` and `_` (italic)
- `MarkdownTag::CodeBlock` for the backtick (`)
- `MarkdownTag::Strikethrough` for `~~`

**Comment by xpe** [comment]
We should also clarify what behavior we want. I suggest the following as a representative example:

1. make selection
2. press `*`
3. one additional `*` is added before and after the selection.

As is explained [here](https://www.markdownguide.org/basic-syntax/), there are three ways to use the `*`:

- once (`*`) for italics (*emphasis*)
- twice (`**`) for bold (**strong**)
- thrice (`***`) for italic and bold (***emphasis and strong***)

Question: What happens when pressing `*` when a selection is already wrapped in `***`? After thinking about it, I would probably recommend removing the `***`.

Question: What happens when `*word*` is selected but `_` is pressed? After thinking about it, I would probably recommend the result be `**word**`.

Note: I predict there will be other edge cases we haven't thought about yet.

**Comment by arv-anshul** [comment]
> 2. What happens when `*word*` is selected but `_` is pressed? After thinking about it, I would probably recommend the result be `**word**`.

I disagree because the user might want the `_*word*_` behaviour instead of `**word**`.

> 4. Would other people here be amenable to including hotkeys e.g. `⌘B` for toggling bold and `⌘I` for toggling italic? This is what [Obsidian](https://forum.obsidian.md/t/make-the-hotkey-for-bold-and-italic-behave-like-in-document-processors-word-google-docs/32239/13) does.

This would be bad for Zed editor users; for example, I uses `⌘B` for sidebar toggling. This might be good fit for Obsidian as it specially for markdown editing.

Although, we may provide a config to set the keyboard shortcut for this.

---

> [!TIP]
>
> We can't just take features from Obsidian like editors as they are majorly focused on Markdown editing.

**Comment by xpe** [comment]
> I disagree because the user might want the _*word*_ behaviour instead of **word**.

Interesting. These are uncommon and fairly weird. After reading https://spec.commonmark.org/0.28/#example-438 and https://spec.commonmark.org/0.28/#example-440, I agree with you:
* `*_foo_*` ➡ `<em><em>foo</em></em>`
* `_*foo*_` ➡ `<em><em>foo</em></em>`

**Comment by xpe** [comment]
>> Would other people here be amenable to including hotkeys e.g. ⌘B for toggling bold and ⌘I for toggling italic?

> This would be bad for Zed editor users; for example, I uses ⌘B for sidebar toggling. ... Although, we may provide a config to set the keyboard shortcut for this.

Good points. I like the idea of having bindings for (1) toggling bold and (2) toggling italic, even if there isn't a keymap to it. (I hope I'm using the terms "binding" and "keymap" in the same way as the Zed project.)

**Comment by xpe** [comment]
<img width="609" alt="Image" src="https://github.com/user-attachments/assets/48b6a411-cb23-4ec6-8ddd-d6efb4ac3b8c" />

⬆ I happen to agree with that -- I just didn't think of it when I was writing my earlier comment -- but I've never seen this style used in a conversation. I see it often it books -- the premise being the author is an expert and the reader is reading the book and looking for advice.

The dynamic is different in a GitHub issue between we're random people on the internet. Using "tip" in this way can rub people the wrong way.

**Comment by arv-anshul** [comment]
Hi, I've reproduced this and the editor only able to tackle the `**` situation not the `__` and `==`. So I think this issue need to kept opened till all the situations being tackled.

### Environment

```
Zed: v0.213.3 (Zed) 
OS: macOS 26.1
Memory: 16 GiB
Architecture: aarch64
```
